Step2

FineRoot1253 opened the pull request.

FineRoot1253 committed:
docs(/): modify README.md

----

## README.md

modify docs

FineRoot1253 committed:
feat(course): add Sessions in Course

----

## Course.java

add sessions

----

## CourseTest.java

add create()

----

## Session.java

add Session()

----

## Sessions.java

add Sessions()

----

## SessionsTest.java

add create()

----

## SessionTest.java

add create()

----

## README.md

modify docs

FineRoot1253 committed:
feat(course): add paymentType, price, enrollment count,
 limit, status

----

## Session.java

add sessionStatus
add sessionPaymentType
add price
add countOfEnrollments
add limitOfEnrollments

----

## SessionEnrollment.java

add isNotEmpty()
add getCountOfEnrollments()
add getLimits()

----

## SessionEnrollmentTest.java

add create()

----

## SessionPayment.java

add isPaid()
add getType()
add getAmount()

----

## SessionPaymentTest.java

add create()

----

## SessionPaymentType.java

add PAID
add FREE

----

## SessionPaymentTypeTest.java

add matching()

----

## SessionStatus.java

add READY
add ENROLLING
add DONE

----

## SessionStatusTest.java

add matching()

----

## SessionTest.java

add create_throw_exception_startDate_endDate()
add create_throw_exception_sessionPaymentType_price()

----

##
README.md

modify docs

FineRoot1253 committed:
feat(course): add NsUsers, enroll features, coverimage feature

----

## Course.java

add enroll()

----

## CoverImage.java

----

## CoverImageTest.java

add create()
add create_ratio_exception()
add create_width_height_exception()

----

## Duration.java

----

## DurationTest.java

add create()
add create_throw_exception_startDate_endDate()

----

## Enrollment.java

modify SessionEnrollment -> Enrollment
modify countOfEnrollments -> users
modify isNotEmpty()
modify isFull()
modify validateCounts()
add enroll()
delete getCountOfEnrollments()
delete getLimits()

----

## EnrollmentTest.java

add create()
add enroll()
add enroll_exception()

----

## ImageExtension.java

add GIF
add JPG
add JPEG
add PNG
add SVG
add from()

----

## NsUsers.java

add add()
add isFull()
add isGreater()
add isNotEmpty()

----

## NsUsersTest.java

add create()
add add()
add isFull()
add isGreater()
add isNotEmpty()

----

## Session.java

modify sessionPaymentType, price -> sessionPayment
modify startDate, endDate -> duration
modify countOfEnrollments, limitOfEnrollments -> enrollment
delete validateDate()
delete validatePrice()
modify Session()
modify validateCountOfEnrollments()
add enroll()
add isSameId()

----

## SessionEnrollmentTest.java

delete file

----

## SessionPayment.java

add validatePrice()
add isSameAmountOfPay()
delete getType()
delete getAmount()

----

## SessionPaymentTest.java

modify package name

----

## Sessions.java

add enroll()

----

## SessionTest.java

modify create()
modify create_throw_exception_sessionPaymentType_price()
delete create_throw_exception_startDate_endDate()

----

## README.md

modify docs

FineRoot1253 committed:
feat(course): add ExceptionMessage, add enroll feature in service

----

## Course.java

modify enroll()

----

## CourseService.java

add enroll()

----

## CoverImage.java

add ExceptionMessage

----

## Duration.java

add ExceptionMessage

----

## Enrollment.java

modify limits type
delete isNotEmpty()
add ExceptionMessage

----

## EnrollmentTest.java

modify enroll()
modify enroll_exception()

----

## ExceptionMessage.java

add COVER_IMAGE_RATIO
add COVER_IMAGE_BYTE_SIZE
add COVER_IMAGE_LEAST_SIZE
add DURATION_RANGE
add ENROLLMENT_SIZE
add IMAGE_EXTENSION_NOT_FOUND_TYPE
add SESSIONS_NOT_FOUND_SESSION
add SESSION_NOT_MATCH_AMOUNT_OF_PAY
add SESSION_STATUS_NOT_ENROLLING
add SESSION_PAYMENT_NEED_PRICE

----

## ImageExtension.java

add ExceptionMessage

----

## NsUserLimit.java

add isLessThan()
add isFull()

----

## NsUsers.java

modify isFull()
modify isGreater()

----

## NsUsersTest.java

modify isFull()
modify isGreater()

----

## Session.java

delete users
modify Session()
delete validateCountOfEnrollments()
modify enroll()
add ExceptionMessage

----

## SessionPayment.java

add ExceptionMessage

----

## Sessions.java

modify enroll()
add ExceptionMessage

----

## SessionTest.java

modify create_throw_exception_sessionPaymentType_price()

catsbi commented on README.md:
```diff
@@ -25,4 +25,26 @@
+## Step 2
+### Requirements
+
+- [x] 과정(Course)는 기수 단위로 운영된다.
```
꼼꼼한 체크리스트 👍

catsbi commented on src/main/java/nextstep/courses/domain/CoverImage.java:
```diff
@@ -0,0 +1,38 @@
+        this.name = name;
+        String[] splitName = name.split("\\.");
+        this.extension = ImageExtension.from(splitName[splitName.length - 1]);
+        validateByteSize(byteSize);
+        this.byteSize = byteSize;
+        validateWidthHeight(width, height);
```
유효성 검증은 항상 로직의 최대한 최상단에 있는걸 권장드려요. 불필요한 변수할당이나 로직을 수행하지 않도록요 ㅎ

catsbi commented on src/main/java/nextstep/courses/domain/CoverImage.java:
```diff
@@ -0,0 +1,38 @@
+    }
+
+    private void validateByteSize(Long byteSize) {
+        if(byteSize > 1000L){
```
1000L이 무엇을 의미하는지 알 수 있게 상수로 이름을 지어주면 어떨까요?

catsbi commented on src/main/java/nextstep/courses/domain/ImageExtension.java:
```diff
@@ -0,0 +1,22 @@
+    }
+
+    public static ImageExtension from(String extension){
+        return Arrays.stream(ImageExtension.values()).filter(e -> e.name.equals(extension)).findFirst().orElseThrow(
+                ()->new IllegalArgumentException(ExceptionMessage.IMAGE_EXTENSION_NOT_FOUND_TYPE.getMessage()));
```
values로 이름 매칭을 한다면, 매번 매칭시마다 Stream 순회를 해야하는데, 캐싱을 활용해보면 어떨까싶네요!

catsbi commented on src/main/java/nextstep/courses/domain/NsUserLimit.java:
```diff
@@ -0,0 +1,19 @@
+package nextstep.courses.domain;
+
+public class NsUserLimit {
+    private final Integer count;
```
명확한 이유(ex: generic)가 있는게 아니라면 primitive 타입 사용을 권장드립니다.

catsbi commented on src/main/java/nextstep/courses/domain/Session.java:
```diff
@@ -0,0 +1,52 @@
+    private Duration duration;
+    private SessionStatus sessionStatus;
+    private CoverImage coverImage;
+    private LocalDateTime createdAt;
+    private LocalDateTime updatedAt;
```
공통적으로 엔티티에 들어가는 정보들은 super type class를 만들어서 관리해도 좋을 것 같네요!

catsbi commented on src/main/java/nextstep/courses/domain/Sessions.java:
```diff
@@ -0,0 +1,19 @@
+import nextstep.users.domain.NsUser;
+
+public class Sessions {
+    private final List<Session> sessionList;
```
collection 타입의 인스턴스변수는 필드초기화를 해주는것도 괜찮습니다. NPE에서 좀 더 안전해질 수 있어서요

catsbi commented on src/main/java/nextstep/courses/service/CourseService.java:
```diff
@@ -0,0 +1,14 @@
+
+import nextstep.users.domain.NsUser;
+
+public class CourseService {
```
사용하는 곳도 테스트하는곳도 존재하지 않네요?

catsbi commented on src/test/java/nextstep/courses/domain/CoverImageTest.java:
```diff
@@ -0,0 +1,33 @@
+    }
+
+    @Test
+    @DisplayName("CoverImage 생성시, 너비, 높이 비율이 3:2가 아니면 예외 발생")
```
예외는 발생한다기보단 던진다(throw)는 표현이 적절합니다.

catsbi commented on src/test/java/nextstep/courses/domain/CoverImageTest.java:
```diff
@@ -0,0 +1,33 @@
+
+class CoverImageTest {
+    @Test
+    @DisplayName("SessionCoverImage")
```
DisplayName 작성에 좀 더 신경을써보면 좋을 것 같아요. 
실제로 거대 엔터프라이즈 프로젝트를 진행하면 테스트케이스가 수백/수천개가 생성되는데, 이 테스트들을 모두 코드를 보면서 분석할순 없기 때문에 DisplayName으로 작성된 내용이 콘솔에 작성된 것을 보고 파악하는데, 이 내용이 가독성이 낮을 경우 업무 효율이 낮아질 수 밖에 없겠죠

catsbi reviewed (changes requested):
안녕하세요 준근님 2단계 미션 잘 진행해주셨는데, 몇 가지 피드백 반영이 필요해보여서 코멘트 남겨놨습니다! 
더하여 현재 테스트가 좀 부족한부분들(CourseService, Enrollment, Course, Sessions, Session, SessionPaymentType, ..)이 있으니 이 부분들에 대한 테스트가 작성되면 좋을 것 같네요. 어떤식으로 테스트를 작성할지 뭘 테스트해야할지 파악하기 힘드시다면 다음 포스팅을 참고해보시면 좋을 것 같습니다! 

참고: https://catsbi.oopy.io/7c084479-c9d0-44a1-acb9-f6b43a19e332

FineRoot1253 committed:
refactor(course): refactor feedback

----

## AnswerTest.java

modify display name

----

## CoverImageTest.java

modify display name

----

## DurationTest.java

modify display name

----

## EnrollmentTest.java

modify display name

----

## Sessions.java

modify Sessions()

----

## SessionsTest.java

modify create()

FineRoot1253 committed:
refactor(course): refactor feedback

----

## AuditInfo.java

add AuditInfo()

----

## Course.java

delete createdAt
delete updatedAt
modify Course()

----

## CourseService.java

add CourseService()

----

## CourseServiceTest.java

add setUp()
add enroll()

----

## CoverImage.java

modify CoverImage()

----

## ImageExtension.java

add extensionCacheMap
modify from()

----

## NsUser.java

add equals()
add hashcode()

----

## NsUserLimit.java

modify count

----

## Session.java

delete createdAt
delete updatedAt
modify Session()

----

## SessionTest.java

modify create_throw_exception_sessionPaymentType_price()
modify display name

catsbi commented on src/main/java/nextstep/courses/domain/AuditInfo.java:
```diff
@@ -0,0 +1,14 @@
+
+import java.time.LocalDateTime;
+
+public abstract class AuditInfo {
```
추상클래스로 공통 필드들을 분리하셨네요 👍

catsbi commented on src/main/java/nextstep/courses/domain/CoverImage.java:
```diff
@@ -10,18 +11,18 @@ public class CoverImage {
 
     private void validateByteSize(Long byteSize) {
-        if(byteSize > 1000L){
+        if(byteSize > LIMIT_BYTE_SIZE){
```
매직리터럴 상수 분리 👍

catsbi commented on src/main/java/nextstep/courses/domain/ImageExtension.java:
```diff
@@ -9,14 +12,20 @@ public enum ImageExtension {
     SVG("svg");
 
-    final String name;
+    private static final Map<String, ImageExtension> extensionCacheMap = Arrays.stream(ImageExtension.values()).collect(
```
캐싱도 잘 해주셨네요!

catsbi reviewed (approved):
준근님 피드백 반영 너무 잘 해주셨어요 💯 
머지하겠습니다. 다음 단계 진행해주세요!
